### DISCUSSION TITLE
Multirate extended Kalman Filter implementation

### DISCUSSION THREAD

**Discussion by jonathanayalaleon** [do-mpc/do-mpc#134] [Q&A]
Hello, 
I am trying to implement a **MEFK**, what I am trying to do is to create a new MHE every time step (n_horizon =1 equivalent to EKF), every time I create a EKF I need to specify the model, the measurements I have available, and the P_v matrix. The model because when I define the model I define which measurements are available, sometimes I have one measurement sometimes I have more than one available. 
My approach to check  that the previous implementation works, was to create a EKF everytime but assuming I always have the same amount of measurements available, this is equivalent to have a single EKF, so I should get exactly the same estimation as using a single EKF( this wouldn't be multirate). Moreover, for the measurements I created a function with mhe.get_y_template(), I want to emphasis this works for the case we have the same amount of measurements for every step. 
For passing information about the previous estimation/optimization, to the new EKF, after creating the new EKF I simply update mhe.opt_x_num with the previous information. the issue is when I want to implement this for the **multirate** case, because I have ten measurements in iteration 1, but in iteration 2 I have one measurement available. so I can not simply update *mhe.opt_x_num* because it has different dimensions, I mean I can update it but the optimizer shows me the following error:

```
RuntimeError: .../casadi/core/function_internal.hpp:1257: Input 0 (x0) has mismatching shape. Got 99-by-1. Allowed   dimensions, in general, are:
- The input dimension N-by-M (here 90-by-1)
- A scalar, i.e. 1-by-1
- M-by-N if N=1 or M=1 (i.e. a transposed vector)
- N-by-M1 if K*M1=M for some K (argument repeated horizontally)
- N-by-P*M, indicating evaluation with multiple arguments (P must be a multiple of 1 for consistency with previous inputs)
```
How would you recommend me to pass the data from one "EKF' to the next one, taking into account that the measurements are not the same for all time steps?
Thank you very much in advance. 
Best regards

**Comment by 4flixt** [comment]
Hi Jonathan,

thank you for your question. I am not entirely sure what you are trying to achieve here: You want to use the MHE as an EKF?
This is a bit confusing to me. Could you elaborate?

Best
4flixt

**Reply by jonathanayalaleon** [reply]
Thanks for replying @4flixt. 
yes, you are right I want to use the MHE as an EKF, the thing is that every time I have measurements available I need to update the P_x matrix, using the Kalman way, that way I think I would implement the multirate EKF, do you have any advices about that? 

Regards 
jonathan

**Reply by 4flixt** [reply]
My advice would be to use an EKF in that case. What do you expect to improve by using the MHE in this way?
